nav2_amcl/src/amcl_node.cpp

```diff
@@ -845,18 +845,18 @@ AmclNode::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
       // p.covariance[6*5+5] = hyps[max_weight_hyp].pf_pose_cov.m[2][2];
       p.pose.covariance[6 * 5 + 5] = set->cov.m[2][2];
 
-      /*
-         printf("cov:\n");
-         for(int i=0; i<6; i++)
-         {
-         for(int j=0; j<6; j++)
-         printf("%6.3f ", p.covariance[6*i+j]);
-         puts("");
-         }
-       */
+      float temp = 0.0;
+      for (int i = 0; i < static_cast<int>(p.pose.covariance.size()); i++) {
+        temp += p.pose.covariance[i];
+      }
+      temp += p.pose.pose.position.x + p.pose.pose.position.y;
+      if (!std::isnan(temp)) {
+        RCLCPP_INFO(get_logger(), "AmclNode publishing pose");
+        pose_pub_->publish(p);
+      } else {
+        RCLCPP_WARN(get_logger(), "ignoring NAN pose");
```

Comment:
If there's a NaN pose/covariance in the pipe, that's really bad and probably isn't going to go away in 1 or 2 cycles, I'd make this more strongly worded and explaining what has happened. If someone sees this there's probably something very wrong with their setup and there are failing sections of safety-critical software.

Reply:
Yes I agree, once there is a NaN the filter is diverged and it probably won't go away without resetting the filter and/or sensors.
When we enable life-cycle node, if NaN is detected we can bring AMCL node to un-configured state, check sensor values, and reset. If all fails then we can print an error message and shutdown.  But for now maybe we can just print ROS Error with "AMCL is diverged and pose is not available!"  Your thoughts?

Reply:
`AMCL covariance or pose is NaN, likely due to an invalid configuration! Pose is not available!` or so

Reply:
When I was seeing this NAN it was due to a bug in the upstream data coming from the TB3, a classic "garbage in" problem. Is it possible here to verify the inputs before calculating the pose and covariance so that we reduce the likelihood of this error ever happening and points the user to the source of the problem?

Reply:
That would be good. This was added here and checking in Navigation I dont see any warnings, personally, I think a crash would be an acceptable output here if a NaN gets introduced. That's an upstream issue, but a warning when you get "garbage in" would aid in debugging why its crashing so its not wrongly blamed on AMCL.
Ironically enough, that's the same type of warning I added to `robot_localization` about a year ago because I was inputting garbage sensor covariance matrices.

Reply:
@mkhansen-intel  I can add another method in addition to this to also check for NaN for input data.

Reply:
@mhpanah - that would be great. Can you check if the input pose values are greater than the size of the map? In the TB3 case, the values were not NAN but they were like 1e30+ due to the bug. Maybe an upper bound of 10x the boundaries of the map? I'm not sure what exactly makes sense here but It seems some sanity check could be done.

Reply:
@mkhansen-intel @SteveMacenski Validating sensors data shouldn't really be AMCL's task. For example if we validate Odom data here and odom data is used elsewhere, then the other module should also implement a way to validate odom data. 

If NaN gets introduced due to invalid configuration or faulty sensor measurements, we should put the AMCL into 1) inactive state, 2) reset AMCL, 3) check for NaN, 4) repeat if it fails, 5) shutdown if problem persists.

We can do the above when life-cycle is available, however, for now not publishing AMCL pose and provide WARN should be okay temporarily.

Reply:
I'm curious how you were that far off the map?

Reply:
@SteveMacenski - it was due to this bug, now fixed. I know it 'shouldn't' happen again, but it's always a good idea to have some checks on your data input where possible.